Integrate `StorageKafka2` to `system.kafka_consumers`

antaljanosbenjamin opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Integrate `StorageKafka2` to `system.kafka_consumers`.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

antaljanosbenjamin committed:
Introduce KeeperHandlingConsumer

antaljanosbenjamin committed:
Improve test

antaljanosbenjamin committed:
Store committed offset only in one place

Before it was stored in both `committed_topic_partition_offsets` and in the locks. From now on the single source of truth is the locks.

antaljanosbenjamin committed:
Protect the lock infos with a mutex

antaljanosbenjamin committed:
Add support for `system.kafka_consumers`

The summary of this big commit in short:
 - `KafkaConsumer`: tiny changes, mostly related to moving the parts of statistic that can be reused in `KafkaConsumer2` to `StorageKafkaUtils`
 - `KafkaConsumer2`: added accounting of some statistics, only the ones that are tied closely with `cppkafka::Consumer`. Some statistics related things had to be added, because only `KafkaConsumer2` doesn't communicate (and it shouldn't) the number of fetched, but not consumed yet message, and providing more up-to-date current offsets was easier by implementing it in this class. Furthermore, because in a future commit we have make sure unused `cppkafka::Consumer`s are closed even when the storage exists, I added those capabilities already to ensure that my plan is feasible and make sense.
 - `KeeperHandlingConsumer`: Added substantial amount of docs and also the parts of statistics that is less closely tied to `cppkafka::Consumer`. I had to add some of the statistics, because the `KafkaConsumer2` is not aware of intent sizes (and shouldn't be).
  - `test_batch_fast.py`: updated `test_kafka_producer_consumer_separate_settings` to trigger the creation of consumers in `StorageKafka2` with the new approach and made `test_system_kafka_consumers` work with the new table engine.
 - In the rest of the files I made only smaller changes to make them suit the new approach

antaljanosbenjamin committed:
Log offset too when saving intent to make `test_intent_sizes.py::test_good_intent_size` pass

antaljanosbenjamin committed:
Remove unused forward declaration

antaljanosbenjamin committed:
Formatting

antaljanosbenjamin committed:
Handle `KafkaConsumersInUse` metric in `StorageKafka2`

antaljanosbenjamin committed:
Tidy-up TODOs

antaljanosbenjamin committed:
Merge branch 'master' into kafka2-system-kafka-consumers

pamarcos commented on src/Storages/Kafka/KeeperHandlingConsumer.h:
```diff
@@ -0,0 +1,196 @@
+
+    struct LockedTopicPartitionInfo
+    {
+        zkutil::EphemeralNodeHolderPtr lock;
```
This seems unused. Not sure if you just meant to have it here because that's the way this lock works. But if it's really protecting something, it might be useful to add TSA annotations to show what's protecting. Feel free to check out `UniqueLock.h` to see how to make a TSA-compatible wrapper.

pamarcos commented on src/Storages/Kafka/KeeperHandlingConsumer.cpp:
```diff
@@ -0,0 +1,618 @@
+    ++poll_count;
+
+    topic_partition_index_to_consume_from = (topic_partition_index_to_consume_from + 1) % assigned_topic_partitions.size();
+    const auto & topic_partition = assigned_topic_partitions[topic_partition_index_to_consume_from];
```
```suggestion
    const auto & topic_partition = assigned_topic_partitions[topic_partition_index_to_consume_from];
```

antaljanosbenjamin replied:
What was the original idea?

pamarcos reviewed (commented):
Wow, this was a huuuuge PR! :rocket: 
I've added quite a few comments. None of them are blockers.

I'm finding a hard time understanding how you added the integration of `StorageKafka2` to `system.kafka_consumer` but yet I haven't seen any explicit test that ensures that it's populated as it should when using `StorageKafka2` instead of `StorageKafka`.

antaljanosbenjamin commented on src/Storages/Kafka/KafkaConsumer2.cpp:
```diff
@@ -59,7 +49,44 @@ KafkaConsumer2::KafkaConsumer2(
+
+void KafkaConsumer2::createConsumer(cppkafka::Configuration consumer_config)
+{
+    chassert(!consumer.get());
```
I think you are right. Now I am thinking about it second time, doesn't make sense to use `chassert`, because it will make the debugging worse.

My original thought was the logic is super simple (don't call `createConsumer` if `hasConsumer` returns `true`), so I assume we will never run into this apart from in the middle of the development.

antaljanosbenjamin commented on tests/integration/test_storage_kafka/test_batch_fast.py:
```diff
@@ -3082,70 +3085,82 @@ def test_block_based_formats_1(kafka_cluster, create_query_generator):
 
-
-def test_system_kafka_consumers(kafka_cluster):
+@pytest.mark.parametrize(
+    "create_query_generator, consumer_id_length, num_rebalance_assignments",
+    [(k.generate_old_create_table_query, 67, 1), (k.generate_new_create_table_query, 0, 0)],
+)
+def test_system_kafka_consumers(kafka_cluster, create_query_generator, consumer_id_length, num_rebalance_assignments):
```
@pamarcos The test is here. The trick is `generate_new_create_table_query` will generate a create query that will result in creating `StorageKafka2`.

antaljanosbenjamin commented on src/Storages/Kafka/KafkaConsumer2.cpp:
```diff
@@ -85,41 +112,43 @@ bool KafkaConsumer2::polledDataUnusable(const TopicPartition & topic_partition)
 {
     resetIfStopped();
 
+    // TODO: `polledDataUnusable` should be renamed to `clearMessagesIfNotSameTopicPartition` and it should be okay to continue without returning nullptr.
```
To stop introducing even more changes. As the code was here before, I thought it is okay to leave a TODO and fix it later.

antaljanosbenjamin commented on tests/integration/test_storage_kafka/test_batch_slow_6.py:
```diff
@@ -285,10 +287,12 @@ def produce():
             # Waiting for test.kafka_consumerX to start consume ...
             instance.wait_for_log_line(log_line.format(table_name))
 
+            logging.debug(instance.query("SELECT count(), uniqExact(key), max(key) + 1 FROM test.destination"))
```
I think it can be useful in the future to detect when some messages got duplicated, after which consumer.

antaljanosbenjamin commented on src/Storages/Kafka/StorageKafka2.cpp:
```diff
@@ -1482,15 +1198,16 @@ zkutil::ZooKeeperPtr StorageKafka2::getZooKeeperIfTableShutDown() const
+    switch (reason)
 {
-    auto topic_partition_name = fmt::format("{}_{}.{}", topic_partition.topic, topic_partition.partition_id, lock_file_name);
-    return fs_keeper_path / "topic_partition_locks" / topic_partition_name;
+        case KeeperHandlingConsumer::CannotPollReason::NoPartitions:
+            return StallReason::NoPartitions;
+        case KeeperHandlingConsumer::CannotPollReason::NoMetadata:
+            return StallReason::NoMetadata;
+        case KeeperHandlingConsumer::CannotPollReason::KeeperSessionEnded:
+            return StallReason::KeeperSessionEnded;
+}
```
The idea is the streaming thread can stall for more reasons than the keeper handling consumer cannot poll messages:
```
    enum class StallReason : uint8_t
    {
        NoMetadata,
        KeeperSessionEnded,
        NoPartitions,
        NoMessages,
        ConsumerNotAvailable,
    };
```
vs
```
    enum class CannotPollReason : uint8_t
    {
        NoMetadata,
        KeeperSessionEnded,
        NoPartitions,
    };
```

I reworked this to use `StallKind` because the only important information is whether we want to stall the stream for a short period or a longer period:
```
    enum class StallKind : uint8_t
    {
        ShortStall,
        LongStall,
    };
```

antaljanosbenjamin commented on src/Storages/Kafka/StorageKafka2.cpp:
```diff
@@ -1359,63 +1045,102 @@ std::optional<StorageKafka2::StallReason> StorageKafka2::streamToViews(size_t id
-std::optional<size_t> StorageKafka2::streamFromConsumer(ConsumerAndAssignmentInfo & consumer_info)
+void StorageKafka2::cleanConsumers()
+{
+    /// We need to clear the cppkafka::Consumer separately from KafkaConsumer2, since cppkafka::Consumer holds a weak_ptr to the KafkaConsumer2 (for logging and stat callback).
+    /// So if we destroy cppkafka::Consumer in KafkaConsumer2 destructor, then due to librdkafka will call the logging again from destructor, it will lead to a deadlock.
+    /// Maybe we could do this in the destructor of KeeperHandlingConsumer, thus avoid this not obvious logic here, but this version is "battle tested" by our CI as we have
+    /// the very similar, if not the same approach in the old StorageKafka. Let's go with this now, later on we can improve it.
```
I rewrapped the lines to 120. If I missed anything (I focused only in `StorageKafka2/KeeperHandlingConsumer/KafkaConsumer2`), please let me know.

antaljanosbenjamin committed:
Fix minor PR comments

antaljanosbenjamin committed:
Make `test_kafka_unavailable` less flaky

Reading from a kafka engine while a materialized view is attached is not allowed. Thus first do the read, then attach the materialized view to make sure the direct read won't fail because of the attach materialized view.

antaljanosbenjamin committed:
Define missing error code

pamarcos reviewed (approved):
<img width="245" height="170" alt="image" src="https://github.com/user-attachments/assets/9559b03b-d0ba-45fd-ac04-0a84520d8447" />
